### ISSUE TITLE
Train Loss Remaining at 0 in Single-Class Classification Training with single_cls=True

### ISSUE THREAD

**Issue by HosseinMoazzam** [ultralytics/ultralytics#8560]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I want to train YOLOv8 for a single-class classification problem. I placed my images in a single folder within the train, validation, and test folders. Then, I executed the training command:

model = YOLO('model/yolov8m-cls.pt')
Result = model.train(data='C:\Data\Bottle', single_cls=True, plots=True, batch=64, imgsz=640, epochs=100)

Upon completion of training, the loss remains at zero from epoch 0 to the final epoch (in this case, 100), indicating that the model did not learn anything. Post-training, the model consistently classifies all images (include images belong that class or any other classes) into the same class with a confidence score of 1.

If you could provide guidance on where I might be mistaken.

![1](https://github.com/ultralytics/ultralytics/assets/73931196/97dcdf70-c4b7-4fb2-813a-8fd208ad17ea)
![2](https://github.com/ultralytics/ultralytics/assets/73931196/1be42ea8-faa7-4ca1-bbbf-e80f2c374631)
![results](https://github.com/ultralytics/ultralytics/assets/73931196/c187617e-aea5-464d-b68a-1235b2181b10)

**Comment by Jordan-Pierce** [comment]
You're training an image classifier on a single class... When you start training, the loss (error) will immediately go to 0.0 because the final output layer only has one option to choose from: your class. Meaning it will always be correct in it's guess, leading it to be 100% "correct" all of the time - 100% accurate with a 0.0 loss. 

The problem is not the model or your parameters, it's that you're training with a single class. See [here](https://www.quora.com/How-would-you-train-a-network-using-only-one-class-of-images) for more details.

**Comment by Jordan-Pierce** [comment]
@HosseinMoazzam, you said in your original post that you're training a "YOLOv8 for a single-class classification problem". How many registered classes are in your dataset (`data.yaml`, `nc`)?

**Comment by HosseinMoazzam** [comment]
@Jordan-Pierce, I have only one class and one directory for the images of this class that I will use for training.

**Comment by Jordan-Pierce** [comment]
@HosseinMoazzam so yeah, whatever your end-goal is, if you fine-tune a model on a dataset that consists of a single class, or you set `single_cls` to `True`, your trained model be 100% accurate and have a loss of 0.0, but will always predict every image it receives as that one class. Which is not really useful in it of itself. 

I'm assuming that that is not your goal, so you might need to re-evaluate how you setup your data for the task you're trying to accomplish.

**Comment by HosseinMoazzam** [comment]
@Jordan-Pierce, I want to detect the outliers or anomalies via single-class classification.

**Comment by Jordan-Pierce** [comment]
@HosseinMoazzam then you need to create a dataset containing two classes: normal, and anomalies, and train a model for a binary-classification task.

**Comment by HosseinMoazzam** [comment]
@Jordan-Pierce, so what is the application of single-class classification, and what should I do when I don't have enough images for the anomalies or I don't have any images for that?

**Comment by Jordan-Pierce** [comment]
Single-class classification usually means binary-classification, [but not always, for example, here they use a one-class SVM](https://machinelearningmastery.com/one-class-classification-algorithms/#:~:text=One%2DClass%20Classification%2C%20or%20OCC,and%20potential%20outliers%20to%20appear.). For this model (YOLOv8) you'll probably need to use two classes (normal vs anomaly).[ Here's another person who ran into the same issue as you, and what they did](https://stackoverflow.com/questions/57309958/one-class-classification-using-keras-and-python).

"what should I do when I don't have enough images for the anomalies or I don't have any images for that?" - literally the reason why anomaly detection is difficult.

One thing that you could try is training the model on all of your data as you're doing, and then, use it as a feature extractor which you then use to train a clustering algorithm. For any new samples you want to make predictions on, you pass it through the model, get the feature vector, and apply the clustering model on it and based on some threshold value, you label it as normal or anomaly. 

I'd also look at the probability or logits being output from your trained YOLO model for samples that are normal vs anomalies to see if there's some stark difference, in which case that will help you determine where to go next.

**Comment by glenn-jocher** [comment]
@Jordan-Pierce, single-class classification can indeed be tricky, especially in the context of anomaly detection where you might not have examples of the "anomaly" class. 🤔 In such cases, traditional binary classification (normal vs anomaly) might not be directly applicable due to the lack of anomaly samples.

One approach, as you mentioned, is to focus on one-class classification, where the model learns from a single class and identifies anything that doesn't fit this class as an anomaly. This is a common strategy in anomaly detection when you have a well-defined "normal" state but anomalies are rare or not well-defined. 

Since YOLOv8 and similar models are primarily designed for multi-class object detection, adapting them for one-class classification or anomaly detection might require some creativity. One idea is to use the model to extract features from your "normal" images, and then apply a one-class classifier, like a One-Class SVM, on these features to distinguish normal from abnormal. This approach leverages the powerful feature extraction capabilities of YOLOv8 while employing a specialized algorithm for anomaly detection.

Here's a simplified example of how you might set this up:

```python
from ultralytics import YOLO
from sklearn.svm import OneClassSVM
import numpy as np

# Load your YOLOv8 model
model = YOLO('yolov8n.pt')

# Assume X_train contains your "normal" images loaded as numpy arrays
# Extract features using YOLOv8
features = [model.predict(image, features=True) for image in X_train]

# Train a One-Class SVM on these features
clf = OneClassSVM(gamma='auto').fit(features)

# For a new sample
new_sample_feature = model.predict(new_sample, features=True)
prediction = clf.predict([new_sample_feature])

# If prediction is -1, it's an anomaly; if 1, it's normal
```

This is a high-level approach and would need adjustments based on your specific dataset and requirements. Also, exploring the output probabilities or logits from YOLOv8 for normal vs potential anomalies could provide insights into setting a threshold for anomaly detection.

Remember, anomaly detection is inherently challenging, and there's often no one-size-fits-all solution. Experimentation and iteration will be key. Good luck! 🍀
